### PR TITLE
Added Support for 1.16.3 & 1.16.4

### PR DESCRIPTION
**Pull request by QuentiumYT**
image

After seeing this message, I changed the versions and added new files for 1.16.3 and 1.16.4 support (as LOOHP did)
I tried to compile but it seems I am missing something `The POM for org.bukkit:craftbukkit:jar:1.16.4-R0.1-SNAPSHOT is missing, no dependency information available`. Tried to compile without my modifs didn't changed anything.
The spigot API have both new versions btw (https://hub.spigotmc.org/nexus/content/repositories/snapshots/org/spigotmc/spigot-api/)
I hope it will work on your side :)

### CODE REVIEW

**Comment by Qveshn** [comment]
Hi I think you have made a mistake
version `1.16.4` not equals namespace  `net.minecraft.server.v1_16_R4`
R4 - I think "R" means something like "revision" (IMHO)

Here is a list of the correct mapping version to nms prefix:
1.16.1 = v1_16_R1
1.16.2 = v1_16_R2
1.16.3 = v1_16_R2
1.16.4 = v1_16_R3

But nevermind. I have almost completed the update to 1.16.4.
I will publish new version in some hours

**Comment by QuentiumYT** [comment]
Hi Qveshn,
Oh well, I didn't knew that, to be honest, I never coded in java haha, just wondering if I could help you with the new 1.16.4 release :p
I learned a lot while trying to compile this project haha, it was very interesting!

This is weird `1.16.3 = v1_16_R2` it makes no sense but okay ^^, where did you find this informations just to know?

Thanks for keeping this plugin updated!

**Comment by Qveshn** [comment]
> This is weird `1.16.3 = v1_16_R2` it makes no sense but okay ^^, where did you find this informations just to know?

Here: https://www.spigotmc.org/wiki/spigot-nms-and-minecraft-versions-1-16/

That page has not been updated since Sep 29, 2020. Therefore there is no info about 1.16.4. I hope it will be updated soon.
But now here is full list of versions:


Version | Bukkit Version | NMS Version | Jenkins build
-- | -- | -- | --
1.8 | 1.8-R0.1-SNAPSHOT | 1_8_R1 | 1.8
1.8.3 | 1.8.3-R0.1-SNAPSHOT | 1_8_R2 | 422
1.8.4 | 1.8.8-R0.1-SNAPSHOT | 1_8_R3 | 582
1.8.5 | 1.8.8-R0.1-SNAPSHOT | 1_8_R3 | 582
1.8.6 | 1.8.8-R0.1-SNAPSHOT | 1_8_R3 | 582
1.8.7 | 1.8.8-R0.1-SNAPSHOT | 1_8_R3 | 582
1.8.8 | 1.8.8-R0.1-SNAPSHOT | 1_8_R3 | 582
1.9 | 1.9-R0.1-SNAPSHOT | 1_9_R1 | 765
1.9.2 | 1.9.2-R0.1-SNAPSHOT | 1_9_R1 | 818
1.9.4 | 1.9.4-R0.1-SNAPSHOT | 1_9_R2 | 849
1.10 | 1.10.2-R0.1-SNAPSHOT | 1_10_R1 | 986
1.10.2 | 1.10.2-R0.1-SNAPSHOT | 1_10_R1 | 986
1.11 | 1.11-R0.1-SNAPSHOT | 1_11_R1 | 1110
1.11.1 | 1.11.2-R0.1-SNAPSHOT | 1_11_R1 | 1251
1.11.2 | 1.11.2-R0.1-SNAPSHOT | 1_11_R1 | 1251
1.12 | 1.12-R0.1-SNAPSHOT | 1_12_R1 | 1371
1.12.1 | 1.12.1-R0.1-SNAPSHOT | 1_12_R1 | 1399
1.12.2 | 1.12.2-R0.1-SNAPSHOT | 1_12_R1 | 1573-j
1.13-pre7 | 1.13-pre7-R0.1-SNAPSHOT | 1_13_R1 | 1654
1.13 | 1.13-R0.1-SNAPSHOT | 1_13_R1 | 1807
1.13.1 | 1.13.1-R0.1-SNAPSHOT | 1_13_R2 | 1915
1.13.2 | 1.13.2-R0.1-SNAPSHOT | 1_13_R2 | 2148-c
1.14-pre5 | 1.14-pre5-SNAPSHOT | 1_14_R1 | 2176
1.14 | 1.14-R0.1-SNAPSHOT | 1_14_R1 | 2294
1.14.1 | 1.14.1-R0.1-SNAPSHOT | 1_14_R1 | 2331
1.14.2 | 1.14.2-R0.1-SNAPSHOT | 1_14_R1 | 2366
1.14.3-pre4 | 1.14.3-pre4-SNAPSHOT | 1_14_R1 | 2372
1.14.3 | 1.14.3-R0.1-SNAPSHOT | 1_14_R1 | 2405
1.14.4 | 1.14.4-R0.1-SNAPSHOT | 1_14_R1 | 2502-b
1.15 | 1.15-R0.1-SNAPSHOT | 1_15_R1 | 2520
1.15.1 | 1.15.1-R0.1-SNAPSHOT | 1_15_R1 | 2574
1.15.2 | 1.15.2-R0.1-SNAPSHOT | 1_15_R1 | 2703
1.16.1 | 1.16.1-R0.1-SNAPSHOT | 1_16_R1 | 2850
1.16.2 | 1.16.2-R0.1-SNAPSHOT | 1_16_R2 | 2908
1.16.3 | 1.16.3-R0.1-SNAPSHOT | 1_16_R2 | 2935
1.16.4 | 1.16.4-R0.1-SNAPSHOT | 1_16_R3 | 2943

**Comment by QuentiumYT** [comment]
Alright, I tried to compile the updated git but seems to be a problem on my side...
Anyway, thanks you and have a nice day!

TLDR: I will close this PR :)

**Comment by Qveshn** [comment]
1. clone repository to your local folder
2. install java jdk if missing
3. install maven if missing (I am using maven built in IntelliJ IDEA as plugin)
4. Run script at the root of LightAPI folder:

```
set JAVA_HOME=<path to java jdk>
<path to mvn.cmd or mvn.bat> clean package
```

Example on my side (Windows cmd-file):
```
@echo off
set JAVA_HOME=c:\Program Files\Java\jdk1.8.0_212
"c:\Program Files\JetBrains\IntelliJ IDEA Community Edition 2019.1.3\plugins\maven\lib\maven3\bin\mvn.cmd" clean package
```

Also you can download binary jar file
here: https://github.com/Qveshn/LightAPI/releases/tag/3.4.6
and here: https://www.spigotmc.org/resources/lightapi-fork.48247/
and even here: https://oss.sonatype.org/content/repositories/snapshots/io/github/qveshn/LightAPI-fork/ 😃 

Also you can use maven dependency in you project pom file (if you want to develop your own plugin with dependence on LightAPI):
```
    <repositories>
        ...
        <repository>
            <id>sonatype</id>
            <url>https://oss.sonatype.org/content/groups/public/</url>
        </repository>
    </repositories>

    <dependencies>
         ...
        <dependency>
            <groupId>io.github.qveshn</groupId>
            <artifactId>LightAPI-fork</artifactId>
            <version>3.4.6-SNAPSHOT</version>
        </dependency>
    </dependencies>
```

**Comment by QuentiumYT** [comment]
I am using a linux server to compile your plugin and did this:
```
$ java -version
java version "1.8.0_251"
Java(TM) SE Runtime Environment (build 1.8.0_251-b08)
Java HotSpot(TM) 64-Bit Server VM (build 25.251-b08, mixed mode)
$ wget https://github.com/Qveshn/LightAPI/archive/3.4.6.zip
$ unzip 3.4.6.zip
$ cd LightAPI-3.4.6/
$ ../apache-maven-3.6.3/bin/mvn clean package
```
(Maven is on the main user folder and the path is correct)

But while compiling, it still show this error:
```
[INFO] common 3.3.0 ....................................... SUCCESS [  6.200 s]
[INFO] v1_16_R3 3.3.0 ..................................... FAILURE [  0.024 s]
etc.....
[ERROR] Failed to execute goal on project v1_16_R3: Could not resolve dependencies for project ru.beykerykt.lightapi.nms:v1_16_R3:jar:3.3.0: Could not find artifact org.bukkit:craftbukkit:jar:1.16.4-R0.1-SNAPSHOT -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/DependencyResolutionException
[ERROR] 
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <args> -rf :v1_16_R3
```
And this is the same issue I got since yesterday when I tried to compile.

**Comment by Qveshn** [comment]
Oops... I forgot one main thing... 😟 

LightAPI project depends on the craftbukkit servers implementations for next versions:
1.8.8 1.9 1.9.4 1.10.2 1.11 1.12.2 1.13 1.13.2 1.14.4 1.15 1.16.1 1.16.3 1.16.4
Unfortunately, these servers are not available on the Internet due to the distribution policy.
But you can easily create them yourself.
To do it you need:
1. Download utility BuildTools.jar to build craftbukkit/spigot servers:
https://hub.spigotmc.org/jenkins/job/BuildTools/
2. Place BuildTools.jar in any temporary empty folder and run there next commands:
```
java -jar BuildTools.jar --compile craftbukkit --rev 1.8.8
java -jar BuildTools.jar --compile craftbukkit --rev 1.9
java -jar BuildTools.jar --compile craftbukkit --rev 1.9.4
java -jar BuildTools.jar --compile craftbukkit --rev 1.10.2
java -jar BuildTools.jar --compile craftbukkit --rev 1.11
java -jar BuildTools.jar --compile craftbukkit --rev 1.12.2
java -jar BuildTools.jar --compile craftbukkit --rev 1.13
java -jar BuildTools.jar --compile craftbukkit --rev 1.13.2
java -jar BuildTools.jar --compile craftbukkit --rev 1.14.4
java -jar BuildTools.jar --compile craftbukkit --rev 1.15
java -jar BuildTools.jar --compile craftbukkit --rev 1.16.1
java -jar BuildTools.jar --compile craftbukkit --rev 1.16.3
java -jar BuildTools.jar --compile craftbukkit --rev 1.16.4
```
This will compile all needed servers and also will place them in local maven cache.
This will take quite a while. But be patient. It took me about 1 hour for all servers.
There may be some errors in console when building, but as for me. I ignore them. The main thing is not to stop the process.

After all will be done you will see in this folder 13 craftbukkit servers for needed versions (jar-files).
All their copies BuildTools.jar should have also installed in local maven cache (I hope).
You can check it. But I do not know where is local maven cache on linux. May be in `var` folder or `home` folder?
I think the path should contain `.../.m2/repository/org/bukkit/craftbukkit`
Now you can forgot the temporary folder with BuildTools.jar. You can also delete this folder if you do not need to run these servers or something else to do with them.

After all this, you can repeat the LightAPI  build process:
```
mvn clean package
```
😃

**Comment by QuentiumYT** [comment]
Oh yeah! I was sure it was because of craftbukkit new system with BuildTools.

I have seen spigot publishing their builds [here](https://hub.spigotmc.org/nexus/content/repositories/snapshots/org/spigotmc/spigot-api/) but if you use craftbukkit, we have to compile ourselves :p

And in Linux, it's in the main user folder under `.m2/repository/org/` and yeah, craftbukkit does not shows up in here for me ^^

Anyway, thank you for all the documentation and details you've provided, just wow :) (you should copy it in a CONTRIBUTE.MD perhaps)
Such an amazing job, I hope next release I can provide a better PR with all this knowledge haha

All the best for this project and have a nice day!

**Comment by Qveshn** [comment]
https://hub.spigotmc.org/nexus/content/repositories/snapshots/org/spigotmc/spigot-api/
It is not server. It is API for developing spigot plugins (not full server). This is Spigot api.

And
https://hub.spigotmc.org/nexus/content/repositories/snapshots/org/bukkit/bukkit/
This is bukkit and again, this is not server. It is API for developing bukkit plugins (not full server)

Summarizing:
* Bukkit API:
File name looks like `bukkit-... .jar`
Documentation: https://hub.spigotmc.org/javadocs/bukkit/index.html (the same as for spigot API)
Maven public repository: https://hub.spigotmc.org/nexus/content/repositories/snapshots/org/bukkit/bukkit/
Local maven cache folder: `.../.m2/repository/org/bukkit/bukkit/`
* Bukkit SERVER:
File name looks like `craftbukkit-... .jar`
Documentation: NO
Maven public repository: NO (distribution policy)
Local maven cache folder: `.../.m2/repository/org/bukkit/craftbukkit/`

* Spigot API:
File name looks like `spigot-api-... .jar`
Documentation: https://hub.spigotmc.org/javadocs/spigot/index.html
Maven public repository: https://hub.spigotmc.org/nexus/content/repositories/snapshots/org/spigotmc/spigot-api/
Local maven cache folder: `.../.m2/repository/org/spigotmc/spigot-api/`
* Spigot SERVER:
File name looks like `spigot-... .jar`
Documentation: NO
Maven public repository: NO (distribution policy)
Local maven cache folder: `.../.m2/repository/org/spigotmc/spigot/`

The main difference between API and SERVER implementation, that API contains mostly only interfaces.
And SERVER contains classes that implement these interfaces.

Usually only API is required for plugin development, but LightAPI uses internal SERVER classes. Therefore, it depends on SERVER implementation (API not enough).
